File: Codigo/clases.py
import random

class Nodo:
    def __init__(self, ID):
        self.ID = ID
        self.color = None
        self.vecinos = []
        self.numvecinos = 0
    
    def colorear(self, nuevo_color):
        self.color = nuevo_color
    
    def añadir_vecino(self, vecino):
        if vecino not in self.vecinos:
            self.vecinos.append(vecino)
            self.numvecinos = len(self.vecinos)
    
    def eliminar_vecino(self, vecino):
        if vecino in self.vecinos:
            self.vecinos.remove(vecino)
            self.numvecinos = len(self.vecinos)

    def contar_conflictos(self):
        contador=0
        for vecino in self.vecinos:
            if vecino.color == self.color:
                contador+=1
        return contador

    def obtener_colores_validos(self, colores):
        for vecino in self.vecinos:
            if vecino.color in colores:
                colores.remove(vecino.color)
        return colores

class Arista:
    def __init__(self, nodo1, nodo2):
        self.nodo1 = nodo1
        self.nodo2 = nodo2
        self.conflicto = False
  
class Grafo:
    def __init__(self):
        self.nodos = []
        self.aristas = []
        self.colores = ["Azul", "Amarillo", "Morado"]
        
    def agregar_nodo(self, nodo):
        if nodo not in self.nodos:
            self.nodos.append(nodo)

    def eliminar_nodo(self, nodo):
        if nodo in self.nodos:
            self.nodos.remove(nodo)
            for vecino in nodo.vecinos:
                vecino.eliminar_vecino(nodo)
        for arista in self.aristas:
            if nodo == arista.nodo1 or nodo == arista.nodo2:
                self.aristas.remove(arista)

    def agregar_arista(self, nodo1, nodo2):
        if nodo1 in self.nodos and nodo2 in self.nodos:
            arista = Arista(nodo1, nodo2)
            if arista not in self.aristas and Arista(nodo2, nodo1) not in self.aristas:
                nodo1.añadir_vecino(nodo2)
                nodo2.añadir_vecino(nodo1)
                self.aristas.append(arista)
                
    def eliminar_arista(self, nodo1, nodo2):
        arista = Arista(nodo1, nodo2)
        if arista in self.aristas:
            nodo1.eliminar_vecino(nodo2)
            nodo2.eliminar_vecino(nodo1)
            self.aristas.remove(arista)
            
    def obtener_conexiones(self):
        conexiones = []
        for arista in self.aristas:
            conexiones.append(((arista.nodo1.ID, arista.nodo2.ID), arista.conflicto))
        return conexiones
    def obtener_nodos(self):
        ids = []
        for nodo in self.nodos:
            ids.append((nodo.ID, nodo.color))
        return ids
    
    def colorar_aleatoriamente(self):
        for nodo in self.nodos:
            nodo.colorear(random.choice(self.colores))
        self.validar_aristas()
        
    def validar_aristas(self):
        for arista in self.aristas:
            if arista.nodo1.color == arista.nodo2.color and arista.nodo1.color is not None:
                arista.conflicto = True
            else:
                arista.conflicto = False
                
    def total_conflictos(self):
        cont=0
        for arista in self.aristas:
            if arista.conflicto:
                cont+=1
        return cont
    
    def validar_coloreo(self):
        if self.total_conflictos()>0:
            return False
        return True
    
    def obtener_colores(self):
        mapa_colores = {}
        for nodo in self.nodos:
            mapa_colores[nodo.ID] = nodo.color
        return mapa_colores
    
    def recolorear_nodos(self, diccionario_colores):
        for nodo in self.nodos:
            if nodo.ID in diccionario_colores:
                nodo.colorear(diccionario_colores[nodo.ID])
        self.validar_aristas()  

    def mostrar_grafo(self):
        for nodo in self.nodos:
            print(f"Nodo {nodo.ID}: Color={nodo.color}, Vecinos={[vecino.ID for vecino in nodo.vecinos]} Conflictos={nodo.contar_conflictos()}")
        print(f"Estadisticas del grafo: Total Nodos={len(self.nodos)}, Total Aristas={len(self.aristas)}, Colores Disponibles={self.colores}, Conflictos Totales={self.total_conflictos()}")
        print(f"Porcentaje de Conflictos: {self.total_conflictos()/len(self.aristas)*100:.4f}%")

File: Codigo/src/Clases.ts
/**
 * Clase Nodo
 *
 * Representa un vértice del grafo con:
 * - id: identificador numérico
 * - color: nombre del color asignado o null si no tiene color
 * - vecinos: lista de referencias a nodos adyacentes
 * - numvecinos: contador redundante del número de vecinos (mantenido manualmente)
 */
export class Nodo {
  id: number;
  color: string | null;
  vecinos: Nodo[];
  numvecinos: number;
  constructor(id: number) {
    this.id = id;
    this.color = null;
    this.vecinos = [];
    this.numvecinos = 0;
  }

    /**
   * Asigna un color al nodo.
   */
  colorear(nuevo_color: string) {
    this.color = nuevo_color;
  }

    /**
   * Registra una conexión con otro nodo.
   *
   * Evita duplicados.
   */
  añadir_vecino(vecino: Nodo) {
    if (!this.vecinos.includes(vecino)) {
      this.vecinos.push(vecino);
      this.numvecinos++;
    }
  }

    /**
   * Elimina una relación de vecindad entre nodos.
   */
  eliminar_vecino(vecino: Nodo) {
    const indice = this.vecinos.indexOf(vecino);
    if (indice !== -1) {
      this.vecinos.splice(indice, 1);
      this.numvecinos--;
    }
  }

    /**
   * Cuenta cuántos vecinos tienen el mismo color que el nodo.
   */
  contar_conflictos() {
    let contador = 0;
    for (const vecino of this.vecinos) {
      if (vecino.color === this.color) contador++;
    }
    return contador;
  }
}

/**
 * Clase Arista
 *
 * Representa una arista no dirigida entre dos nodos.
 * - nodo1, nodo2: referencias a instancias de Nodo.
 * - conflicto: bandera booleana que indica si ambos extremos comparten color.
 */
export class Arista {
  nodo1: Nodo;
  nodo2: Nodo;
  conflicto: boolean;
  constructor(nodo1: Nodo, nodo2: Nodo) {
    this.nodo1 = nodo1;
    this.nodo2 = nodo2;
    this.conflicto = false;
  }
}

/**
 * Clase Grafo
 *
 * Estructura que contiene nodos y aristas y ofrece utilidades para:
 * - añadir/eliminar nodos y aristas
 * - coloreo aleatorio del grafo
 * - validación de aristas/conflictos
 * - obtención de estadísticas y listados
 *
 * Observaciones:
 * - `colores` define la paleta disponible (por defecto 3 colores).
 * - Varias operaciones mutan el grafo (colorear, recolorear, añadir/aristas, etc.).
 */
export class Grafo {
  nodos: Nodo[];
  aristas: Arista[];
  colores: string[];
  constructor() {
    this.nodos = [];
    this.aristas = [];
    this.colores = ["Azul", "Amarillo", "Morado"];
  }

    /**
   * Agrega un nodo al grafo.
   */
  añadir_nodo(nodo: Nodo) {
    this.nodos.push(nodo);
  }

    /**
   * Conecta dos nodos mediante una arista.
   *
   * Actualiza vecindades.
   */
  añadir_arista(nodo1: Nodo, nodo2: Nodo) {
    const arista = new Arista(nodo1, nodo2);
    this.aristas.push(arista);
    nodo1.añadir_vecino(nodo2);
    nodo2.añadir_vecino(nodo1);
  }

    /**
   * Elimina una arista del grafo y limpia referencias internas.
   */
  eliminar_arista(arista: Arista) {
    const indice = this.aristas.indexOf(arista);
    if (indice !== -1) {
      this.aristas.splice(indice, 1);
      arista.nodo1.eliminar_vecino(arista.nodo2);
      arista.nodo2.eliminar_vecino(arista.nodo1);
    }
  }

    /**
   * Retorna las conexiones del grafo con indicador de conflicto.
   */
  obtener_conexiones() {
    this.validar_aristas();
    const conexiones = [];
    for (const arista of this.aristas) {
      conexiones.push([arista.nodo1.id, arista.nodo2.id, arista.conflicto]);
    }
    return conexiones;
  }

    /**
   * Retorna los nodos con su estado de color.
   */
  obtener_nodos() {
    const ids = [];
    for (const nodo of this.nodos) {
      ids.push([nodo.id, nodo.color]);
    }
    return ids;
  }

    /**
   * Asigna colores aleatorios a todos los nodos.
   */
  colorear_grafo() {
    for (const nodo of this.nodos) {
      const indice_Aleatorio = Math.floor(Math.random() * this.colores.length);
      const color_aleatorio = this.colores[indice_Aleatorio];
      nodo.colorear(color_aleatorio);
    }
    this.validar_aristas();
  }

    /**
   * Marca conflictos en cada arista según los colores.
   */
  validar_aristas() {
    for (const arista of this.aristas) {
      if (arista.nodo1.color === arista.nodo2.color) {
        arista.conflicto = true;
      } else {
        arista.conflicto = false;
      }
    }
  }

    /**
   * Retorna la cantidad total de aristas en conflicto.
   */
  total_conflictos() {
    let total = 0;
    for (const arista of this.aristas) {
      if (arista.conflicto) total++;
    }
    return total;
  }

    /**
   * Determina si el grafo es una coloración válida.
   */
  validar_coloracion() {
    if (this.total_conflictos() === 0) {
      return true;
    } else {
      return false;
    }
  }

    /**
   * Aplica un mapa externo de colores a los nodos existentes.
   */
  recolorear_nodos(diccionario_colores: { [key: number]: string | null }) {
    for (const nodo of this.nodos) {
      if (nodo.id in diccionario_colores) {
        const color = diccionario_colores[nodo.id];
        if (color !== null) {
          nodo.colorear(color);
        }
      }
    }
    this.validar_aristas();
  }

    /**
   * Devuelve el mapa actual de colores del grafo.
   */
  obtener_colores() {
    const diccionario_colores: { [key: number]: string | null } = {};
    for (const nodo of this.nodos) {
      diccionario_colores[nodo.id] = nodo.color;
    }
    return diccionario_colores;
  }

    /**
   * Calcula métricas globales del grafo.
   *
   * Incluye:
   * - Conflictos.
   * - Densidad.
   * - Número de colores.
   * - Distribución de colores.
   */
  obtener_estadisticas() {
    const total_nodos = this.nodos.length;
    const total_aristas = this.aristas.length;
    const total_conflictos = this.total_conflictos();
    const es_valido = this.validar_coloracion();
    const porcentaje_conflictos =
      total_aristas > 0
        ? ((total_conflictos / total_aristas) * 100).toFixed(2)
        : "0.00";

    // Contar nodos por color
    const nodos_por_color: { [color: string]: number } = {};
    for (const nodo of this.nodos) {
      if (nodo.color) {
        nodos_por_color[nodo.color] = (nodos_por_color[nodo.color] || 0) + 1;
      }
    }

    // Contar nodos con conflictos
    let nodos_con_conflictos = 0;
    for (const nodo of this.nodos) {
      if (nodo.contar_conflictos() > 0) {
        nodos_con_conflictos++;
      }
    }

    // Densidad del grafo
    const max_aristas = (total_nodos * (total_nodos - 1)) / 2;
    const densidad =
      max_aristas > 0
        ? ((total_aristas / max_aristas) * 100).toFixed(2)
        : "0.00";

    return {
      total_nodos,
      total_aristas,
      total_conflictos,
      es_valido,
      porcentaje_conflictos: parseFloat(porcentaje_conflictos),
      k_colores: this.colores.length,
      colores_disponibles: this.colores,
      nodos_por_color,
      nodos_con_conflictos,
      densidad_grafo: parseFloat(densidad),
      aristas_validas: total_aristas - total_conflictos,
    };
  }

    /**
   * Lista los nodos que presentan conflictos.
   */
  obtener_nodos_conflictivos() {
    const nodos_conflictivos = [];
    for (const nodo of this.nodos) {
      const num_conflictos = nodo.contar_conflictos();
      if (num_conflictos > 0) {
        nodos_conflictivos.push({
          id: nodo.id,
          color: nodo.color,
          conflictos: num_conflictos,
        });
      }
    }
    return nodos_conflictivos;
  }
}
